rfc-176-replacing-static.md

```diff
@@ -0,0 +1,127 @@
+
+Longer-term, we could replace the redis call with a similarly-cached call to a [content-store] item for the various banners, which would allow us to make the various banners first-class content items with scheduling support (see also "The Global Banner"). It’s worth noting that we could also replace redis with content-store calls directly in `static`, but either way the change would be done in only one place, so that more desirable solution to handling banner content is neutral regarding whether the display is done in `static` or a gem. To implement banners in `content-store` we could either create a new schema for banners or perhaps implement them as reusable content blocks, if that project becomes realised.
+
+### The Global Banner (move to `govuk_web_banners`)
+
+The current mechanism for the global banner is via code directly in `static`. Adding a banner is a developer task. This allows one slight advantage over the emergency banner, which is that the code can change the banner content at a particular time (eg [removing a voter id banner] when the polls close in an election)
+
+Our proposal is that the global banner ultimately be deployed using the same mechanism as the emergency banner, but with additional information baked into the values put into the redis store. Since redis naturally supports key expiration, the act of removing banners at a certain time would be pretty easy. Banners appearing at a certain time would be more awkward, but not insurmountable. These changes would make it deployable by content specialists as well as developers, and should be able to be implemented with minimal changes to `whitehall`.
+
+In the short-term, just moving the existing global banner code into the gem with the recruitment banners and emergency banner would also simplify ensuring that recruitment and global banners are not shown when an emergency banner is in effect.
+
+Longer-term, maintaining the same mechanism as emergency banners, we could look to make global banners first-class publishing items as well, which would give us standard scheduled publication options.
+
+### Recruitment Banners (move to `govuk_web_banners`)
+
+This isn’t a problem with `static`, but fits the banner theme and provides a neat start to the banner gem. Recruitment banners present a problem in that they often need to be inserted into the body of a page inside the area that `static` normally considers the app’s section (the wrapper). This means they need to be maintained in multiple apps.
+
+Our proposal is that we include the current duplicated recruitment banner config, code, and partial views in the gem. Then we can include the gem in each application and put a single call to it in the relevant app layouts. This means we can keep the configuration in one place, where it’s more easily checked for problems, and a single release of the banner gem will be automatically pulled into each relevant app at the next dependabot merge.
```

Comment:
Yes. I don't think there's another way around that as the recruitment banners are rendered in the body the page rather than then header.

I think this is ok though. It's something we already have with other gems we import, and also similar to the current setup with the frontend apps calling static for the layout from govuk_publishing_components which they already have access too.

While not ideal, I think this is better than the the recruitment banner code being replicated in 4 places. The only other option is to have a separate "recruitment banners" gem, but I think that would only make the location of banner code harder to find.

Reply:
Going to mark this as resolve since there won't be a layouts gem in the current (hopefully final) incarnation of the proposal.